service/storage_service.cc

```diff
@@ -1906,48 +1908,171 @@ future<> storage_service::do_stop_ms() {
-        auto req = node_ops_cmd_request{cmd, uuid, {}, {}, {}};
+
+    future<> abort(node_ops_cmd cmd) noexcept {
+        co_await stop_heartbeat_updater();
+        co_await send_to_all(cmd);
+    }
+
+    future<> abort_on_error(node_ops_cmd cmd, std::exception_ptr ex) noexcept {
+        slogger.error("{}[{}]: operation failed, sync_nodes={}: {}", desc, uuid(), sync_nodes, ex);
         try {
-          co_await coroutine::parallel_for_each(nodes, [this, ops, uuid, &req] (const gms::inet_address& node) {
-            return _messaging.local().send_node_ops_cmd(netw::msg_addr(node), req).then([ops, uuid, node] (node_ops_cmd_response resp) {
-                slogger.debug("{}[{}]: Got heartbeat response from node={}", ops, uuid, node);
-                return make_ready_future<>();
-            });
-          });
+            co_await abort(cmd);
         } catch (...) {
-            auto ep = std::current_exception();
-            slogger.warn("{}[{}]: Failed to send heartbeat: {}", ops, uuid, ep);
+            slogger.warn("{}[{}]: abort failed while handling error, sync_nodes={}: {}. Ignoring", desc, uuid(), sync_nodes, std::current_exception());
+        }
+        co_await coroutine::return_exception_ptr(std::move(ex));
+    }
+
+    future<> send_to_all(node_ops_cmd cmd) {
+        req.cmd = cmd;
+        sstring op_desc = format("{}", cmd);
+        bool is_prepare = op_desc.find("prepare") != sstring::npos;
+        slogger.info("{}[{}]: Started {}, sync_nodes={}", desc, uuid(), op_desc, sync_nodes);
+        co_await coroutine::parallel_for_each(sync_nodes, [&] (const gms::inet_address& node) -> future<> {
+            if (nodes_unknown_verb.contains(node) || nodes_down.contains(node) ||
+                    (nodes_failed.contains(node) && op_desc != "abort")) {
+                // Note that we still send to failed nodes that 
+                co_return;
+            }
+            try {
+                co_await ss._messaging.local().send_node_ops_cmd(netw::msg_addr(node), req);
+                slogger.debug("{}[{}]: Got {} response from node={}", desc, uuid(), op_desc, node);
+            } catch (const seastar::rpc::unknown_verb_error&) {
+                slogger.warn("{}[{}]: Node {} does not support {} verb", desc, uuid(), node, op_desc);
+                nodes_unknown_verb.emplace(node);
+            } catch (const seastar::rpc::closed_error&) {
+                slogger.warn("{}[{}]: Node {} is down for {} verb", desc, uuid(), op_desc, node);
+                nodes_down.emplace(node);
+            } catch (...) {
+                slogger.warn("{}[{}]: Node {} failed {} verb: {}", desc, uuid(), node, op_desc, std::current_exception());
+                nodes_failed.emplace(node);
+            }
+        });
+        sstring msg;
+        sstring sep = "";
+        if (!nodes_failed.empty()) {
+            msg = format("nodes={} failed {} operation", nodes_failed, op_desc);
+            sep = ";";
```

Comment:
Is it better to put the error msgs in a container and join with ";"?  So no more tricks with the sep all the places.

Reply:
okay, can do.

Reply:
In v3